src/asha_bt.cpp

```diff
@@ -1240,58 +399,249 @@ static void sm_packet_handler (uint8_t packet_type, uint16_t channel, uint8_t *p
+                &scan_gatt_event_handler,
+                curr_scan.ha.conn_handle,
+                &curr_scan.ha.chars.asp,
+                GATT_CLIENT_CHARACTERISTICS_CONFIGURATION_NOTIFICATION
+            );
+            break;
+        }
+        break;
+    }
+    case ScanState::SubscribeASPNotification:
+    {
+        switch(hci_event_packet_get_type(packet)) {
+        case GATT_EVENT_QUERY_COMPLETE: 
+        {
+            auto att_status = gatt_event_query_complete_get_att_status(packet);
+            if (att_status != ATT_ERROR_SUCCESS) {
+                LOG_ERROR("Enabling AudioStatusPoint notifications failed with error code: 0x%02x\n"
+                    "Disconnecting\n", att_status);
+                scan_state = ScanState::Disconnecting;
+                gap_disconnect(curr_scan.ha.conn_handle);
+            }
+            LOG_INFO("AudioStatusPoint notification enabled.\n");
+            scan_state = ScanState::Finalizing;
+            finalise_curr_discovery();
+            break;
+        }
+        }
+        break;
+    }
+    }
+}
 
-    sm_event_callback_registration.callback = &sm_packet_handler;
-    sm_add_event_handler(&sm_event_callback_registration);
+static void connected_gatt_event_handler(uint8_t packet_type, uint16_t channel, uint8_t *packet, uint16_t size)
+{
+    switch (hci_event_packet_get_type(packet)) {
+    case GATT_EVENT_NOTIFICATION:
+    {
+        HA& ha = ha_mgr.get_by_conn_handle(gatt_event_notification_get_handle(packet));
+        if (ha) {
+            int8_t status_val = (int8_t)gatt_event_notification_get_value(packet)[0];
+            ha.on_asp_notification(status_val);
+        }
+        break;
+    }
+    case GATT_EVENT_QUERY_COMPLETE:
+    {
+        HA& ha = ha_mgr.get_by_conn_handle(gatt_event_query_complete_get_handle(packet));
+        if (ha) {
+            ha.on_gatt_event_query_complete(gatt_event_query_complete_get_att_status(packet));
+        }
+        break;
+    }
+    default: 
+        break;
+    }
+}
 
-    LOG_INFO("HCI power on.\n");
-    hci_power_control(HCI_POWER_ON);
+/* Once the Read Only Properties and PSM characteristics have been read,
+   and the AudioStatusPoint notification subscribed to, determine what Device 
+   to assign it to, and whether to continue searching for another aid. */
+static void finalise_curr_discovery()
+{
+    // This shouldn't be the case, but best be sure.
+    if (ha_mgr.set_complete()) {
+        LOG_INFO("Already have complete set. Not adding current device.\n");
+        scan_state = ScanState::Complete;
+        return;
+    }
+    curr_scan.ha.state = HA::State::GATTConnected;
+    curr_scan.ha.l2cap_packet_handler = &l2cap_cbm_event_handler;
+    curr_scan.ha.gatt_packet_handler = &connected_gatt_event_handler;
+    auto& ha = ha_mgr.add(curr_scan.ha);
+    if (!ha) {
+        LOG_ERROR("Error adding this device.\n");
+        scan_state = ScanState::Scan;
+        return;
+    }
+    if (ha_mgr.set_complete()) {
```

Comment:
I wouldn't wait until both devices are available to start audio streaming. Just because devices report themselves as part of a pair doesn't mean that the user wants to use both of them. The stream tends to be much more stable with only one device, since it uses half the bandwidth and the devices aren't trying to sync. There are also use cases (such as somebody using a [cross device](https://github.com/thewierdnut/asha_pipewire_sink/issues/11), where stereo streaming is redundant because the hearing devices just mix them together into one ear anyways.

Reply:
Hmm yeah, good point.